packages/@o3r/test-helpers/src/test-environments/create-test-environment-angular.ts

```diff
@@ -0,0 +1,149 @@
+    globalFolderPath: process.cwd(),
+    registry: 'http://127.0.0.1:4873',
+    replaceExisting: true,
+    ...inputOptions
+  };
+  const packageManager = getPackageManager();
+  const appFolderPath = path.join(options.cwd, options.appDirectory);
+  const locker = new Locker({
+    lockFilePath: path.join(options.cwd, `${options.appDirectory}-ongoing.lock`),
+    lockTimeout: 10 * 60 * 1000
+  });
+  if (options.useLocker && locker.isLocked()) {
+    return locker.waitUntilUnlocked();
+  }
+  if (existsSync(appFolderPath)) {
+    if (!options.replaceExisting) {
+      return;
+    }
+    const packageJson: PackageJson = JSON.parse(readFileSync(path.join(appFolderPath, 'package.json'), {encoding: 'utf8'}));
+    const dependenciesToCheck = [
+      {name: '@angular-devkit/schematics', expected: options.angularVersion, actual: packageJson.dependencies?.['@angular-devkit/schematics'] || 'latest'},
+      {name: '@angular/material', expected: options.materialVersion, actual: packageJson.dependencies?.['@angular/material'] || 'latest'}
+    ];
+    if (dependenciesToCheck.every(({expected, actual}) => !expected || satisfies(expected, actual))) {
+      // No need to regenerate
+      return;
+    }
+  }
+  if (options.useLocker) {
+    locker.lock();
+  }
+
+  const execAppOptions: ExecSyncOptions = {
+    cwd: appFolderPath,
+    stdio: 'inherit',
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    env: {...process.env, NODE_OPTIONS: '', CI: 'true'}
+  };
+
+  // Prepare folder
+  if (existsSync(appFolderPath)) {
+    rmSync(appFolderPath, {recursive: true});
+  }
+
+  // Create app with ng new
+  const createOptions = `--directory=${options.appDirectory} --style=scss --routing --skip-git --package-manager=${packageManager}`;
+  packageManagerCreate(`@angular@${options.angularVersion || ''} ${options.appName} ${createOptions}`,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    {...execAppOptions, cwd: options.cwd});
+
+  setPackagerManagerConfig(options, execAppOptions);
+
+  // Add dependencies
+  const dependenciesToInstall = {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    '@angular-devkit/schematics': options.angularVersion,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    '@angular/pwa': options.angularVersion,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    '@angular/material': options.materialVersion,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    '@angular-devkit/core': options.angularVersion,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    '@schematics/angular': options.angularVersion
+  };
+  packageManagerAdd(`${Object.entries(dependenciesToInstall).map(([depName, version]) => `${depName}@${version || ''}`).join(' ')}`, execAppOptions);
+
+  // Run ng-adds
+  packageManagerExec(`ng add @angular/pwa@${options.angularVersion || ''}`, execAppOptions);
```

Comment:
is `ng add @angular/pwa@` supported (in case of `options.angularVersion === undefined`) or should we use `ng add @angular/pwa@latest` instead

Reply:
I removed the version on the `ng add` since we already set the version on the `yarn add` just before

Reply:
In that case, does `ng add` execute only the `ng-add` schematics or is it trying to upgrade to the latest version also?
If it process to un upgrade, maybe running `ng g @angular/pwa:ng-add` is safer

Reply:
as far as I understand, `ng-add` will check if the currently installed version match the requested version
here, as the version is not specified any installed should match, so it won't try to update it

also, I cannot call `ng g @angular/pwa:ng-add` as it is marked as private/hidden https://github.com/angular/angular-cli/blob/main/packages/angular/pwa/collection.json

Reply:
(Not a good idea from Angular to put is as `private` as today running `ng g ....:ng-add` manually can fix several pnp related issue)